miner/src/block_assembler.rs

```diff
@@ -413,6 +418,192 @@ impl BlockAssembler {
+                if check_tx_skip(&tx_entry) {
+                    continue;
+                }
+
+                if !check_limit_is_ok(&tx_entry) {
+                    consecutive_failed += 1;
+                    // give up
+                    if consecutive_failed > MAX_CONSECUTIVE_FAILED {
+                        break;
+                    }
+                    continue;
+                }
+
+                // find new tx entry
+                entry.replace(tx_entry.to_owned());
+                return;
+            }
+            // set entry to None if iter is end or consecutive failed
+            entry.take();
+        }
+
+        let mut size: usize = 0;
+        let mut cycles: Cycle = 0;
+        let tx_pool = chain_state.tx_pool();
+        let entries = tx_pool.with_sorted_proposed_txs_iter(|iter| {
+            let mut entries: Vec<TxEntry> = Vec::new();
+            // modified entries, after put a tx into block,
+            // the scores of descendants txs should be updated,
+            // these modified entries is stored in modified_entries.
+            // in each loop,
+            // we pick tx from modified_entries and pool to find the best tx to package
+            let mut modified_entries: TxModifiedEntries = TxModifiedEntries::default();
+            // track in block txs
+            let mut in_block_txs: HashSet<ProposalShortId> = HashSet::new();
+            let mut pool_tx = None;
+            let mut modified_tx = None;
+            loop {
+                // 1. choose best tx from pool and modified entries
+                // 2. compare the two txs, package the better one
+                // 3. update modified entries
+                find_tx_entry(
+                    &mut pool_tx,
+                    iter,
+                    size_limit,
+                    cycles_limit,
+                    size,
+                    cycles,
+                    |tx| {
+                        let tx_id = tx.transaction.proposal_short_id();
+                        in_block_txs.contains(&tx_id) || modified_entries.contains_key(&tx_id)
+                    },
+                );
+                modified_entries.with_sorted_by_score_iter(|iter| {
+                    find_tx_entry(
+                        &mut modified_tx,
+                        iter,
+                        size_limit,
+                        cycles_limit,
+                        size,
+                        cycles,
+                        |_| false,
+                    );
+                });
+                // take tx with higher scores
+                let tx_entry = match cmp::max(&mut pool_tx, &mut modified_tx).take() {
+                    Some(entry) => entry,
+                    None => {
+                        // can't find any satisfied tx
+                        break;
+                    }
+                };
+                debug_assert!(!in_block_txs.contains(&tx_entry.transaction.proposal_short_id()));
```

Comment:
If we do have a transaction that has already been proposed, we should skip the following code in the loop.

Reply:
In logic, duplicate txs should be filtered by the above lines.